src/execution_plan/ops/op_delete.c

```diff
@@ -189,50 +186,76 @@ static Record DeleteConsume(OpBase *opBase) {
 
 			SIValue_Free(value);
 		} else if(!(type & T_NULL)) {
-			/* Expression evaluated to a non-graph entity type
-			 * clear pending deletions and raise an exception. */
-			array_clear(op->deleted_nodes);
-			array_clear(op->deleted_edges);
-			// If evaluating the expression allocated any memory, free it.
+			// if evaluating the expression allocated any memory, free it.
 			SIValue_Free(value);
 			// free the Record this operation acted on
 			OpBase_DeleteRecord(r);
 			ErrorCtx_RaiseRuntimeException("Delete type mismatch, expecting either Node or Relationship.");
 			break;
 		}
 	}
-
-	return r;
 }
 
-static OpBase *DeleteClone(const ExecutionPlan *plan, const OpBase *opBase) {
-	ASSERT(opBase->type == OPType_DELETE);
-	OpDelete *op = (OpDelete *)opBase;
-	AR_ExpNode **exps;
-	array_clone_with_cb(exps, op->exps, AR_EXP_Clone);
-	return NewDeleteOp(plan, exps);
+static Record _handoff(OpDelete *op) {
+	return array_pop(op->records);
 }
 
-static OpResult DeleteReset(OpBase *opBase) {
+static Record DeleteConsume(OpBase *opBase) {
 	OpDelete *op = (OpDelete *)opBase;
+	Record r;
 
-	_DeleteEntities(op);
+	// return mode, all data was consumed
+	if(op->records) return _handoff(op);
```

Comment:
Here's a nice Idea, what if once the operation is done pulling records from its child it will update the op's `Consume` function pointer to `_handoff` this way we can eliminate the if condition.

Reply:
then we need to have a reset that return to the correct consume it can be confusing

Reply:
Correct. I don't mind going both ways